File: backend/src/main/java/com/federico/taskforge/service/TaskService.java
package com.federico.taskforge.service;

import com.federico.taskforge.domain.entity.Project;
import com.federico.taskforge.domain.entity.Task;
import com.federico.taskforge.domain.entity.User;
import com.federico.taskforge.domain.enums.TaskPriority;
import com.federico.taskforge.domain.enums.TaskStatus;
import com.federico.taskforge.dto.request.CreateTaskRequest;
import com.federico.taskforge.dto.request.UpdateTaskRequest;
import com.federico.taskforge.dto.request.UpdateTaskStatusRequest;
import com.federico.taskforge.dto.response.TaskResponse;
import com.federico.taskforge.exception.ResourceNotFoundException;
import com.federico.taskforge.repository.ProjectRepository;
import com.federico.taskforge.repository.TaskRepository;
import com.federico.taskforge.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class TaskService {

    private final TaskRepository    taskRepository;
    private final ProjectRepository projectRepository;
    private final UserRepository    userRepository;

    public Page<TaskResponse> findAll(
            Long projectId, TaskStatus status, TaskPriority priority,
            Long assigneeId, Pageable pageable) {

        return taskRepository
            .findAllFiltered(projectId, status, priority, assigneeId, pageable)
            .map(TaskResponse::from);
    }

    public TaskResponse findById(Long id) {
        return TaskResponse.from(getTask(id));
    }

    @Transactional
    public TaskResponse create(CreateTaskRequest request) {
        Project project = projectRepository.findById(request.projectId())
            .orElseThrow(() -> ResourceNotFoundException.of("Project", request.projectId()));

        User assignee = null;
        if (request.assigneeId() != null) {
            assignee = userRepository.findById(request.assigneeId())
                .orElseThrow(() -> ResourceNotFoundException.of("User", request.assigneeId()));
        }

        Task task = Task.builder()
            .title(request.title())
            .description(request.description())
            .priority(request.priority())
            .project(project)
            .assignee(assignee)
            .dueDate(request.dueDate())
            .build();

        return TaskResponse.from(taskRepository.save(task));
    }

    @Transactional
    public TaskResponse update(Long id, UpdateTaskRequest request) {
        Task task = getTask(id);

        if (request.title()       != null) task.setTitle(request.title());
        if (request.description() != null) task.setDescription(request.description());
        if (request.status()      != null) task.setStatus(request.status());
        if (request.priority()    != null) task.setPriority(request.priority());
        if (request.dueDate()     != null) task.setDueDate(request.dueDate());

        if (request.assigneeId() != null) {
            User assignee = userRepository.findById(request.assigneeId())
                .orElseThrow(() -> ResourceNotFoundException.of("User", request.assigneeId()));
            task.setAssignee(assignee);
        }

        return TaskResponse.from(taskRepository.save(task));
    }

    @Transactional
    public TaskResponse updateStatus(Long id, UpdateTaskStatusRequest request) {
        Task task = getTask(id);
        task.setStatus(request.status());
        return TaskResponse.from(taskRepository.save(task));
    }

    @Transactional
    public void delete(Long id) {
        taskRepository.delete(getTask(id));
    }

    private Task getTask(Long id) {
        return taskRepository.findById(id)
            .orElseThrow(() -> ResourceNotFoundException.of("Task", id));
    }
}

File: frontend/src/app/core/services/task.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { Page } from '../../shared/models/api.model';
import {
  Task,
  TaskStatus,
  TaskPriority,
  CreateTaskRequest,
  UpdateTaskRequest,
  UpdateTaskStatusRequest,
} from '../../shared/models/task.model';

export interface TaskFilters {
  projectId?:  number;
  status?:     TaskStatus;
  priority?:   TaskPriority;
  assigneeId?: number;
  page?:       number;
  size?:       number;
}

@Injectable({ providedIn: 'root' })
export class TaskService {
  private readonly api = inject(ApiService);
  private readonly base = '/tasks';

  getAll(filters: TaskFilters = {}): Observable<Page<Task>> {
    const params: Record<string, string | number | boolean> = {};
    if (filters.projectId  != null) params['projectId']  = filters.projectId;
    if (filters.status     != null) params['status']     = filters.status;
    if (filters.priority   != null) params['priority']   = filters.priority;
    if (filters.assigneeId != null) params['assigneeId'] = filters.assigneeId;
    if (filters.page       != null) params['page']       = filters.page;
    if (filters.size       != null) params['size']       = filters.size;
    return this.api.getPaged<Task>(this.base, params);
  }

  getById(id: number): Observable<Task> {
    return this.api.get<Task>(`${this.base}/${id}`);
  }

  create(request: CreateTaskRequest): Observable<Task> {
    return this.api.post<Task>(this.base, request);
  }

  update(id: number, request: UpdateTaskRequest): Observable<Task> {
    return this.api.put<Task>(`${this.base}/${id}`, request);
  }

  updateStatus(id: number, request: UpdateTaskStatusRequest): Observable<Task> {
    return this.api.patch<Task>(`${this.base}/${id}/status`, request);
  }

  delete(id: number): Observable<void> {
    return this.api.delete<void>(`${this.base}/${id}`);
  }
}
